client/src/components/WeatherForm.jsx

```diff
@@ -0,0 +1,87 @@
+import React, { useState } from "react";
+import WeatherCard from "./WeatherCard";
+import {
+  CircularProgress,
+  TextField,
+  Button,
+  Container,
+  Paper,
+} from "@mui/material";
+
+const WeatherForm = () => {
+  const [city, setCity] = useState("");
+  const [weatherData, setWeatherData] = useState("");
+  const [error, setError] = useState(null);
+  const [loading, setLoading] = useState(false);
+
+  const getWeather = async () => {
+    try {
+      setLoading(true);
+      const response = await fetch(
+        `http://localhost:8080/weather?city=${city}`
+      );
+      const data = await response.json();
+
+      if (data.error) {
+        setError(data.error);
+        setWeatherData(null);
+      } else {
+        setWeatherData(data);
+        setError(null);
+      }
+    } catch (error) {
+      console.error("Yikes! Error fetching weather data: ", error);
+      setError("Uhoh! Typo? Please enter a valid city name and try again.");
+      setWeatherData(null);
+    } finally {
+      setLoading(false);
+    }
```

Comment:
The `finally` block is so that whatever code inside of it will run regardless of the result of the `try/catch`. So even if there is an error, the `finally` block will run. I asked ChatGPT if there is a difference between having a `finally` block and just putting `setLoading(false)` outside of the `try/catch` and it says it will achieve the same result. I guess if `finally` is an extra bit of code that doesn't really make a difference here, I can go ahead and delete it.

Reply:
Ok! If `finally` is more readable in this case, I'll replace it.